File: codes/ReverseKGroup25/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        if(head == null || k == 1){
            return head;
        }
        Stack<ListNode> stack = new Stack<ListNode>();
        ListNode dummyhead = new ListNode(0, head);
        ListNode insertPos = dummyhead;
        ListNode curNode = dummyhead.next;
        while(curNode != null){
            while(curNode != null && stack.Count < k){
                stack.Push(curNode);
                curNode = curNode.next;
            }
            if(curNode == null && stack.Count != k)
            {
                while(stack.Count > 1){
                    stack.Pop();
                }
                insertPos.next = stack.Pop();
            }
            else
            {
                while(stack.Count > 0){
                    insertPos.next = stack.Pop();
                    insertPos = insertPos.next;
                    insertPos.next = null;
                }
            }
        }
        return dummyhead.next;
    }
}

File: codes/ReverseKGroup25/Solution.java
package codes.ReverseKGroup25;

import java.util.*;

/*
 * public class ListNode {
 *   int val;
 *   ListNode next = null;
 * }
 */

public class Solution {
    /**
     * 
     * @param head ListNode类 
     * @param k int整型 
     * @return ListNode类
     */
    public ListNode reverseKGroup (ListNode head, int k) {
        // write code here
        ListNode dummy = new ListNode(-1);
        ListNode pre;
        ListNode end;
        ListNode left;
        ListNode right;
        
        
        dummy.next = head;
        end = dummy;
        //left = dummy.next;
        pre = dummy;
        
        while(end != null){
            
            for(int i=0; i < k && end != null; i++) {
                end = end.next;
            }
            
            if(end == null) 
                break;
            
            right = end.next;
            left = pre.next;
            end.next = null;
            
            pre.next = reverse(left);
            left.next = right;
            pre = left;
            //left = left.next;
            end = pre;
        }
        return dummy.next;
        
        
    }
    
    private ListNode reverse (ListNode head) {
        ListNode newHead = new ListNode(-1);
        
        while(head != null) {
            ListNode temp = head.next;
            head.next = newHead;
            newHead = head;
            head = temp;
        }
        
        return newHead;
    }
}
